In [8]:
import pandas as pd

In [9]:
file_url = 'https://raw.githubusercontent.com/sedeba19/Chapter-16/main/data_source/Dataset_processed.cleveland.data.txt'

In [10]:
# Load the data
df_heartd = pd.read_csv(file_url,
                        sep = ',',
                        header= None,
                        na_values= '?')
df_heartd.shape

(303, 14)

In [11]:
df_heartd

0    1    2      3      4    5    6      7    8    9    10   11   12  \
0    63.0  1.0  1.0  145.0  233.0  1.0  2.0  150.0  0.0  2.3  3.0  0.0  6.0   
1    67.0  1.0  4.0  160.0  286.0  0.0  2.0  108.0  1.0  1.5  2.0  3.0  3.0   
2    67.0  1.0  4.0  120.0  229.0  0.0  2.0  129.0  1.0  2.6  2.0  2.0  7.0   
3    37.0  1.0  3.0  130.0  250.0  0.0  0.0  187.0  0.0  3.5  3.0  0.0  3.0   
4    41.0  0.0  2.0  130.0  204.0  0.0  2.0  172.0  0.0  1.4  1.0  0.0  3.0   
..    ...  ...  ...    ...    ...  ...  ...    ...  ...  ...  ...  ...  ...   
298  45.0  1.0  1.0  110.0  264.0  0.0  0.0  132.0  0.0  1.2  2.0  0.0  7.0   
299  68.0  1.0  4.0  144.0  193.0  1.0  0.0  141.0  0.0  3.4  2.0  2.0  7.0   
300  57.0  1.0  4.0  130.0  131.0  0.0  0.0  115.0  1.0  1.2  2.0  1.0  7.0   
301  57.0  0.0  2.0  130.0  236.0  0.0  2.0  174.0  0.0  0.0  2.0  1.0  3.0   
302  38.0  1.0  3.0  138.0  175.0  0.0  0.0  173.0  0.0  0.0  1.0  NaN  3.0   

     13  
0     0  
1     2  
2     1  
3     0  
4     0  
..   ..  
298   1  
299   2  
300   3  
301   1  
302   0  

[303 rows x 14 columns]

In [12]:
df_heartd.loc[:,13].value_counts()

0    164
1     55
2     36
3     35
4     13
Name: 13, dtype: int64

In [13]:
# Change the target values to 1 & 0
df_heartd.loc[df_heartd[13] > 0, 13] = 1

In [14]:
df_heartd.loc[:,13].value_counts()

0    164
1    139
Name: 13, dtype: int64

In [15]:
# Check for missing values
df_heartd.isna().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    4
12    2
13    0
dtype: int64

In [16]:
df_heartd_clean = df_heartd.dropna(axis = 0)
df_heartd_clean.shape

(297, 14)

In [17]:
df_heartd_clean.dtypes.value_counts()

float64    13
int64       1
dtype: int64

In [18]:
# Assign column names
df_heartd_clean.columns = ['age','sex', 'cp', 'trestbps','chol','fbs',
'restec-g','thalach','exang','oldpeak','slope','ca','thal','label']
df_heartd_clean.head()

age  sex   cp  trestbps   chol  fbs  restec-g  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0       2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0       2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0       2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0       0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0       2.0    172.0    0.0      1.4   

   slope   ca  thal  label  
0    3.0  0.0   6.0      0  
1    2.0  3.0   3.0      1  
2    2.0  2.0   7.0      1  
3    3.0  0.0   3.0      0  
4    1.0  0.0   3.0      0

In [19]:
# Create X and y variables
y = df_heartd_clean.pop('label')
y

0      0
1      1
2      1
3      0
4      0
      ..
297    1
298    1
299    1
300    1
301    1
Name: label, Length: 297, dtype: int64

In [20]:
X = df_heartd_clean
X

age  sex   cp  trestbps   chol  fbs  restec-g  thalach  exang  oldpeak  \
0    63.0  1.0  1.0     145.0  233.0  1.0       2.0    150.0    0.0      2.3   
1    67.0  1.0  4.0     160.0  286.0  0.0       2.0    108.0    1.0      1.5   
2    67.0  1.0  4.0     120.0  229.0  0.0       2.0    129.0    1.0      2.6   
3    37.0  1.0  3.0     130.0  250.0  0.0       0.0    187.0    0.0      3.5   
4    41.0  0.0  2.0     130.0  204.0  0.0       2.0    172.0    0.0      1.4   
..    ...  ...  ...       ...    ...  ...       ...      ...    ...      ...   
297  57.0  0.0  4.0     140.0  241.0  0.0       0.0    123.0    1.0      0.2   
298  45.0  1.0  1.0     110.0  264.0  0.0       0.0    132.0    0.0      1.2   
299  68.0  1.0  4.0     144.0  193.0  1.0       0.0    141.0    0.0      3.4   
300  57.0  1.0  4.0     130.0  131.0  0.0       0.0    115.0    1.0      1.2   
301  57.0  0.0  2.0     130.0  236.0  0.0       2.0    174.0    0.0      0.0   

     slope   ca  thal  
0      3.0  0.0   6.0  
1      2.0  3.0   3.0  
2      2.0  2.0   7.0  
3      3.0  0.0   3.0  
4      1.0  0.0   3.0  
..     ...  ...   ...  
297    2.0  0.0   7.0  
298    2.0  0.0   7.0  
299    2.0  2.0   7.0  
300    2.0  1.0   7.0  
301    2.0  1.0   3.0  

[297 rows x 13 columns]

In [21]:
# Split the dataset into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size= 0.3,
                                                    random_state= 123)

In [22]:
X.dtypes.value_counts()

float64    13
dtype: int64

In [23]:
# Create a Processing Engine 
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

num_transformer = Pipeline(steps = [('scaler', StandardScaler())])
num_features = X.select_dtypes(include= 'float').columns

from sklearn.compose import ColumnTransformer
preprocessor = ColumnTransformer(transformers=[('num', num_transformer, num_features)])
preprocessor

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('scaler', StandardScaler())]),
                                 Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restec-g', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal'],
      dtype='object'))])

In [24]:
# Import libraries
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier

# Create a list of classifiers
clas = [KNeighborsClassifier(),
        LogisticRegression(random_state= 123),
        RandomForestClassifier(random_state=123),
        AdaBoostClassifier(random_state= 123)]

# Create a loop through the classifiers to identify the best model
for i in clas:
    modl = Pipeline(steps=[('preprocessor', preprocessor),
                           ('dimred', PCA(10)),
                           ('classifier', i)])
    modl.fit(X_train, y_train)
    print(modl)
    print('Model score: %.2f' % modl.score(X_test, y_test))

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('scaler',
                                                                   StandardScaler())]),
                                                  Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restec-g', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal'],
      dtype='object'))])),
                ('dimred', PCA(n_components=10)),
                ('classifier', KNeighborsClassifier())])
Model score: 0.78
Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('scaler',
                                                                   StandardScaler())]),
                                                  Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restec-g', 'thalach',
       'exang', 'oldpeak', 'slope

In [25]:
# Create a new model Pipeline using Logistic Regression
lr_pipe = Pipeline(steps=[('preprocessor', preprocessor),
                          ('dimred', PCA()),
                          ('classifier', LogisticRegression(random_state=123))])
lr_pipe

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('scaler',
                                                                   StandardScaler())]),
                                                  Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restec-g', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal'],
      dtype='object'))])),
                ('dimred', PCA()),
                ('classifier', LogisticRegression(random_state=123))])

In [26]:
# Declare the parameters as a dictionary
prm_grid =  {'dimred__n_components':[10,11,12,13],
              'classifier__penalty' : ['l1', 'l2'],
              'classifier__C' : [1,3, 5],
              'classifier__solver' : ['liblinear']}

# Instantiate the GridSearchCV
from sklearn.model_selection import GridSearchCV
estimator_gs = GridSearchCV(lr_pipe,
                            cv = 10,
                            param_grid= prm_grid)
estimator_gs

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restec-g', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal'],
      dtype='object'))])),
                                       ('dimred', PCA()),
                                       ('classifier',
                                        LogisticRegression(random_state=123))]),
             param_grid={'classifier__C': [1, 3, 5],
                         'classifier__penalty': ['l1', 'l2'],
                         'classifier__solver': ['liblinear'],
                         'dimred__n_components': [10, 11, 12, 13]})

In [27]:
# Fit the data into model with hyperparameter
estimator_gs.fit(X_train, y_train)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler())]),
                                                                         Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restec-g', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal'],
      dtype='object'))])),
                                       ('dimred', PCA()),
                                       ('classifier',
                                        LogisticRegression(random_state=123))]),
             param_grid={'classifier__C': [1, 3, 5],
                         'classifier__penalty': ['l1', 'l2'],
                         'classifier__solver': ['liblinear'],
                         'dimred__n_components': [10, 11, 12, 13]})

In [28]:
# Print the model's best parameters
estimator_gs.best_params_

{'classifier__C': 1,
 'classifier__penalty': 'l2',
 'classifier__solver': 'liblinear',
 'dimred__n_components': 12}

In [29]:
# Print the model's best score
estimator_gs.best_score_

0.8464285714285713

In [30]:
# Make predictions
pred = estimator_gs.predict(X_test)
pred

array([0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0,
       1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0], dtype=int64)

In [31]:
# Print Classification Report
from sklearn.metrics import classification_report
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.82      0.86      0.84        49
           1       0.82      0.78      0.80        41

    accuracy                           0.82        90
   macro avg       0.82      0.82      0.82        90
weighted avg       0.82      0.82      0.82        90



In [32]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, pred)

array([[42,  7],
       [ 9, 32]], dtype=int64)

From a business perspective, this score of 82% accuracy means that out of the total cases of patient data, we were correctly able to identify 82% of likely cases of customer heart disease. If we look at the recall values of each class, we will see how the model is faring for each class.

From classification report, we can see that a positive with heart disease is 78% while not having a heart disease if 86%. This means that for patients without heart disease, the classifier was more accurately to predict 86% of the available cases in the test set.